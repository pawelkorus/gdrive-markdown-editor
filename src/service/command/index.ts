export * from './CommandsContext';
export * from './Command';
export * from './hooks';
