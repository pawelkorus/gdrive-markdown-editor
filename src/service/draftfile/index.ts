export * from './useDraftFiles'
export * from './useDraftFile'
export * from './types'
