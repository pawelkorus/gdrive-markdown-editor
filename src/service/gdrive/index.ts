import { authorizeInstall as authorizeInstall } from '@app/google'
import { hasPermission, PermissionCheckResult, Permissions } from '@app/google/authorization'
import { useCallback, useEffect, useState } from 'react'

export enum IntegrationStatus {
  LOADING,
  INTEGRATED,
  NOT_INTEGRATED,
}

export function useGdriveIntegration() {
  const [integrationStatus, setIntegrationStatus] = useState(IntegrationStatus.LOADING)

  const integrate = useCallback(async () => {
    await authorizeInstall()
    setIntegrationStatus(IntegrationStatus.INTEGRATED)
  }, [])

  useEffect(() => {
    let currentTimeoutId: string | number | NodeJS.Timeout = null

    function verifyInstallation() {
      if (hasPermission(Permissions.INSTALL) == PermissionCheckResult.PENDING) {
        currentTimeoutId = setTimeout(() => verifyInstallation(), 1000)
      }
      else {
        setIntegrationStatus(
          hasPermission(Permissions.INSTALL) == PermissionCheckResult.GRANTED
            ? IntegrationStatus.INTEGRATED
            : IntegrationStatus.NOT_INTEGRATED,
        )
      }
    }

    verifyInstallation()

    return () => {
      if (currentTimeoutId) {
        clearTimeout(currentTimeoutId)
      }
    }
  }, [])

  return {
    integrationStatus,
    integrate,
  }
}
