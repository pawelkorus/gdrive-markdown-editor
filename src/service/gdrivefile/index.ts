export * from "./GdriveFileContext";
export * from "./hooks";
