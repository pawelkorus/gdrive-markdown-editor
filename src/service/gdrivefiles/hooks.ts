import { useState, useEffect } from 'react'
import { getFileMetadata, FileDetailsWithLink } from '../../google'

export function useGdriveFileMetadata(fileId: string): FileDetailsWithLink | null {
  const [fileMetadata, setFileMetadata] = useState<FileDetailsWithLink | null>(null)

  useEffect(() => {
    const fetchFileMetadata = async () => {
      const metadata = await getFileMetadata(fileId)
      setFileMetadata(metadata)
    }

    fetchFileMetadata()
  }, [fileId])

  return fileMetadata
}
