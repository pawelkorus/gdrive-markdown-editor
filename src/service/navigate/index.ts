import { useNavigate as useReactRouterNavigate, useSearchParams } from 'react-router-dom'

type UseNavigateAPI = {
  navigateToHome: () => void
  navigateToFileView: (params?: NavigateToFileParams) => void
  navigateToFileEdit: (params?: NavigateToFileParams) => void
  navigateToFileSource: (params?: NavigateToFileParams) => void
  navigateToNewFile: (params: NavigateToNewFileParams) => void
}

type NavigateToFileParams = {
  fileId: string
  userId?: string
  resourceKey?: string
}

type NavigateToNewFileParams = {
  folderId: string
}

type ParamsFromURL = {
  paramsFileView: () => NavigateToFileParams
  paramsFileEdit: () => NavigateToFileParams
}

export function useFileParams(): NavigateToFileParams | undefined {
  const [searchParams] = useSearchParams()
  const fileId = searchParams.get('fileId')
  const userId = searchParams.get('userId')
  const resourceKey = searchParams.get('resourceKey')

  if (fileId) {
    return { fileId, userId, resourceKey }
  }
}

export function useParamsFromURL(): ParamsFromURL {
  const [searchParams] = useSearchParams()

  const paramsFileView = (): NavigateToFileParams | undefined => {
    const fileId = searchParams.get('fileId')
    const userId = searchParams.get('userId')
    const resourceKey = searchParams.get('resourceKey')

    if (fileId) {
      return { fileId, userId, resourceKey }
    }
  }

  const paramsFileEdit = (): NavigateToFileParams | undefined => {
    const fileId = searchParams.get('fileId')
    const userId = searchParams.get('userId')
    const resourceKey = searchParams.get('resourceKey')

    if (fileId) {
      return { fileId, userId, resourceKey }
    }
  }

  return {
    paramsFileView,
    paramsFileEdit,
  }
}

export function useNavigateTo(): UseNavigateAPI {
  const navigate = useReactRouterNavigate()
  const {
    paramsFileEdit,
    paramsFileView,
  } = useParamsFromURL()

  const navigateToHome = () => {
    navigate('/')
  }

  const navigateToFileView = (params: NavigateToFileParams | undefined) => {
    if (!params) params = paramsFileView()

    const { fileId, userId, resourceKey } = params
    const searchParams = new URLSearchParams()
    searchParams.set('fileId', fileId)

    if (userId) {
      searchParams.set('userId', userId)
    }
    if (resourceKey) {
      searchParams.set('resourceKey', resourceKey)
    }

    navigate(`/file?${searchParams.toString()}`)
  }

  const navigateToFileEdit = (params: NavigateToFileParams | undefined) => {
    if (!params) params = paramsFileEdit()

    const { fileId, userId, resourceKey } = params
    const searchParams = new URLSearchParams()
    searchParams.set('fileId', fileId)

    if (userId) {
      searchParams.set('userId', userId)
    }
    if (resourceKey) {
      searchParams.set('resourceKey', resourceKey)
    }

    navigate(`/file/edit?${searchParams.toString()}`)
  }

  const navigateToFileSource = (params: NavigateToFileParams | undefined) => {
    if (!params) params = paramsFileView()

    const { fileId, userId, resourceKey } = params
    const searchParams = new URLSearchParams()
    searchParams.set('fileId', fileId)

    if (userId) {
      searchParams.set('userId', userId)
    }
    if (resourceKey) {
      searchParams.set('resourceKey', resourceKey)
    }

    navigate(`/file/source?${searchParams.toString()}`)
  }

  const navigateToNewFile = ({ folderId }: NavigateToNewFileParams) => {
    const searchParams = new URLSearchParams()
    searchParams.set('folderId', folderId)
    const path = `/file/new?${searchParams.toString()}`
    navigate(path)
  }

  return {
    navigateToHome,
    navigateToFileView,
    navigateToFileEdit,
    navigateToFileSource,
    navigateToNewFile,
  }
}
