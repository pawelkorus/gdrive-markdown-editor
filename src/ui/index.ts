export { default as EditorView } from './EditorView'
export { default as ViewerView } from './ViewerView'
export { default as NotificationView } from './NotificationView'
export { default as SourceView } from './SourceView'
export { default as useGlobalCommands } from './useGlobalCommands'
