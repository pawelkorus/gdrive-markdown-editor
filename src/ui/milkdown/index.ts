export { default as MilkdownEditor } from './MilkdownEditor'
export { default as WrapWithProviders } from './WrapWithProviders'
