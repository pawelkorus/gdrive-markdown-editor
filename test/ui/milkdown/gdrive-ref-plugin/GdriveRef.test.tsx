import React from 'react'
import { render, act } from '@testing-library/react'
import { expect, test, vi } from 'vitest'
import { MilkdownEditor, WrapWithProviders } from '@app/ui/milkdown'
import { useNavigateTo } from '@app/service/navigate'
import { useGdriveFileMetadata } from '@app/service/gdrivefiles'

const defaultFileDetails = {
  id: '1',
  name: 'title',
  mimeType: 'text/markdown',
  url: 'http://example.com',
}

vi.mock('@app/service/gdrivefiles', () => {
  return {
    useGdriveFileMetadata: vi.fn(() => defaultFileDetails),
  }
})

vi.mock('@app/service/navigate', () => {
  const navigateToFileView = vi.fn()

  return {
    useNavigateTo: vi.fn(() => {
      return {
        navigateToFileView: navigateToFileView,
      }
    }),
  }
})

test('should render gdrive ref without alt title', async () => {
  const SAMPLE_CONTENT = `:gdrive-ref{src="${defaultFileDetails.id}"}`
  const container = await act(async () => {
    const { container } = render(<WrapWithProviders><MilkdownEditor content={SAMPLE_CONTENT} /></WrapWithProviders>)

    await new Promise<HTMLElement>(r => setTimeout(r, 100))
    return container
  })

  const linkElement = container.querySelector('a')
  expect(linkElement).toHaveAttribute('data-type', 'gdrive-ref')
  expect(linkElement).toHaveAttribute('data-src', '1')
  expect(linkElement).toHaveTextContent(defaultFileDetails.name)
  expect(document.body).toMatchSnapshot()
})

test('should render gdrive ref with alt title', async () => {
  const SAMPLE_LABEL = 'sample label'
  const SAMPLE_CONTENT = `:gdrive-ref[${SAMPLE_LABEL}]{src="${defaultFileDetails.id}"}`
  const container = await act(async () => {
    const { container } = render(<WrapWithProviders><MilkdownEditor content={SAMPLE_CONTENT} /></WrapWithProviders>)

    await new Promise<HTMLElement>(r => setTimeout(r, 100))
    return container
  })

  const linkElement = container.querySelector('a')
  expect(linkElement).toHaveAttribute('data-type', 'gdrive-ref')
  expect(linkElement).toHaveAttribute('data-src', '1')
  expect(linkElement).toHaveTextContent(SAMPLE_LABEL)
  expect(document.body).toMatchSnapshot()
})

test('should navigate to file view when clicked and mime type is markdown', async () => {
  const { navigateToFileView } = useNavigateTo()
  const SAMPLE_CONTENT = `:gdrive-ref{src="${defaultFileDetails.id}"}`
  const container = await act(async () => {
    const { container } = render(<WrapWithProviders><MilkdownEditor content={SAMPLE_CONTENT} /></WrapWithProviders>)

    await new Promise<HTMLElement>(r => setTimeout(r, 100))
    return container
  })

  const linkElement = container.querySelector('a')
  linkElement.click()

  expect(document.body).toMatchSnapshot()
  expect(navigateToFileView).toHaveBeenCalledWith({ fileId: '1' })
})

test('should navigate to file view when clicked and file name extension is md', async () => {
  const fileDetails = { ...defaultFileDetails, name: 'file.md', mimeType: 'text/plain' }
  const { navigateToFileView } = useNavigateTo()
  vi.mocked(useGdriveFileMetadata).mockImplementationOnce(() => fileDetails)

  const SAMPLE_CONTENT = `:gdrive-ref{src="${defaultFileDetails.id}"}`
  const container = await act(async () => {
    const { container } = render(<WrapWithProviders><MilkdownEditor content={SAMPLE_CONTENT} /></WrapWithProviders>)

    await new Promise<HTMLElement>(r => setTimeout(r, 100))
    return container
  })

  const linkElement = container.querySelector('a')
  linkElement.click()

  expect(document.body).toMatchSnapshot()
  expect(navigateToFileView).toHaveBeenCalledWith({ fileId: '1' })
})

test('should not navigate to file view when clicked and not markdown file', async () => {
  const fileDetails = { ...defaultFileDetails, name: 'file', mimeType: 'text/plain' }
  const { navigateToFileView } = useNavigateTo()
  vi.mocked(useGdriveFileMetadata).mockImplementationOnce(() => fileDetails)

  const SAMPLE_CONTENT = `:gdrive-ref{src="${defaultFileDetails.id}"}`
  const container = await act(async () => {
    const { container } = render(<WrapWithProviders><MilkdownEditor content={SAMPLE_CONTENT} /></WrapWithProviders>)

    await new Promise<HTMLElement>(r => setTimeout(r, 100))
    return container
  })

  const linkElement = container.querySelector('a')
  linkElement.click()

  expect(document.body).toMatchSnapshot()
  expect(navigateToFileView).not.toHaveBeenCalledWith({ fileId: '1' })
})
