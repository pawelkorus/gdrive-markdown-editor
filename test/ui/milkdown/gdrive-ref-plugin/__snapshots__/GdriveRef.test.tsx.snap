// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`should navigate to file view when clicked and file name extension is md 1`] = `
<body>
  <script
    defer=""
    src="https://accounts.google.com/gsi/client"
  />
  <script
    defer=""
    src="https://apis.google.com/js/api.js"
  />
  <div>
    <div
      data-milkdown-root="true"
    >
      <div
        class="milkdown"
      >
        <div
          class="ProseMirror editor"
          contenteditable="true"
          role="textbox"
          translate="no"
        >
          <p>
            <span
              data-node-view-root="true"
            >
              <a
                data-src="1"
                data-type="gdrive-ref"
                href="http://example.com"
              >
                file.md
              </a>
            </span>
            <img
              alt=""
              class="ProseMirror-separator"
            />
            <br
              class="ProseMirror-trailingBreak"
            />
          </p>
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`should navigate to file view when clicked and markdown file 1`] = `
<body>
  <script
    defer=""
    src="https://accounts.google.com/gsi/client"
  />
  <script
    defer=""
    src="https://apis.google.com/js/api.js"
  />
  <div>
    <div
      data-milkdown-root="true"
    >
      <div
        class="milkdown"
      >
        <div
          class="ProseMirror editor"
          contenteditable="true"
          role="textbox"
          translate="no"
        >
          <p>
            <span
              data-node-view-root="true"
            >
              <a
                data-src="1"
                data-type="gdrive-ref"
                href="http://example.com"
              >
                title
              </a>
            </span>
            <img
              alt=""
              class="ProseMirror-separator"
            />
            <br
              class="ProseMirror-trailingBreak"
            />
          </p>
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`should navigate to file view when clicked and mime type is markdown 1`] = `
<body>
  <script
    defer=""
    src="https://accounts.google.com/gsi/client"
  />
  <script
    defer=""
    src="https://apis.google.com/js/api.js"
  />
  <div>
    <div
      data-milkdown-root="true"
    >
      <div
        class="milkdown"
      >
        <div
          class="ProseMirror editor"
          contenteditable="true"
          role="textbox"
          translate="no"
        >
          <p>
            <span
              data-node-view-root="true"
            >
              <a
                data-src="1"
                data-type="gdrive-ref"
                href="http://example.com"
              >
                title
              </a>
            </span>
            <img
              alt=""
              class="ProseMirror-separator"
            />
            <br
              class="ProseMirror-trailingBreak"
            />
          </p>
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`should not navigate to file view when clicked and not markdown file 1`] = `
<body>
  <script
    defer=""
    src="https://accounts.google.com/gsi/client"
  />
  <script
    defer=""
    src="https://apis.google.com/js/api.js"
  />
  <div>
    <div
      data-milkdown-root="true"
    >
      <div
        class="milkdown"
      >
        <div
          class="ProseMirror editor"
          contenteditable="true"
          role="textbox"
          translate="no"
        >
          <p>
            <span
              data-node-view-root="true"
            >
              <a
                data-src="1"
                data-type="gdrive-ref"
                href="http://example.com"
              >
                file
              </a>
            </span>
            <img
              alt=""
              class="ProseMirror-separator"
            />
            <br
              class="ProseMirror-trailingBreak"
            />
          </p>
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`should render gdrive ref with alt title 1`] = `
<body>
  <script
    defer=""
    src="https://accounts.google.com/gsi/client"
  />
  <script
    defer=""
    src="https://apis.google.com/js/api.js"
  />
  <div>
    <div
      data-milkdown-root="true"
    >
      <div
        class="milkdown"
      >
        <div
          class="ProseMirror editor"
          contenteditable="true"
          role="textbox"
          translate="no"
        >
          <p>
            <span
              data-node-view-root="true"
            >
              <a
                data-src="1"
                data-type="gdrive-ref"
                href="http://example.com"
              >
                sample label
              </a>
            </span>
            <img
              alt=""
              class="ProseMirror-separator"
            />
            <br
              class="ProseMirror-trailingBreak"
            />
          </p>
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`should render gdrive ref without alt title 1`] = `
<body>
  <script
    defer=""
    src="https://accounts.google.com/gsi/client"
  />
  <script
    defer=""
    src="https://apis.google.com/js/api.js"
  />
  <div>
    <div
      data-milkdown-root="true"
    >
      <div
        class="milkdown"
      >
        <div
          class="ProseMirror editor"
          contenteditable="true"
          role="textbox"
          translate="no"
        >
          <p>
            <span
              data-node-view-root="true"
            >
              <a
                data-src="1"
                data-type="gdrive-ref"
                href="http://example.com"
              >
                title
              </a>
            </span>
            <img
              alt=""
              class="ProseMirror-separator"
            />
            <br
              class="ProseMirror-trailingBreak"
            />
          </p>
        </div>
      </div>
    </div>
  </div>
</body>
`;
